stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -917,17 +931,30 @@ extension ${Self}: BinaryFloatingPoint {
   /// - If `x` is `greatestFiniteMagnitude`, then `x.nextUp` is `infinity`.
   @_inlineable // FIXME(sil-serialize-all)
   public var nextUp: ${Self} {
-    if isNaN { return self }
-    if sign == .minus {
+%if bits == 32 or bits == 64:
+    // Silence signaling NaNs, map -0 to +0.
+    let x = self + 0
 #if arch(arm)
-      // On arm, subnormals are flushed to zero.
-      if (exponentBitPattern == 1 && significandBitPattern == 0) ||
-         (exponentBitPattern == 0 && significandBitPattern != 0) {
-        return ${Self}(sign: .minus,
-          exponentBitPattern: 0,
-          significandBitPattern: 0)
+    // On arm, skip positive subnormal values.
+    if _slowPath(x.bitPattern & (-${Self}.infinity).bitPattern == 0) {
+      return .leastNormalMagnitude
+    }
```

Comment:
The way this is implemented for `arm` worries me. I would do:
~~~~
if (x == 0) {
  return .leastNonzeroMagnitude
}
~~~~
There's no formal logic in IEEE 754 for flushing subnormals, so there's no real spec to go off of, but I prefer to think of them as non-canonical encodings of zero. Implementing them as I describe here does that, and also has the happy side-effect of working correctly for free if someone makes an arm32 platform where subnormals are supported.

Reply:
I like that interpretation. It does change existing behavior though, as [negative subnormal values currently have nextUp evaluating to -0 on arm32](https://github.com/apple/swift/pull/15021/files#diff-1c8541c68839b7021db7e4ab68135c3eL925). But if you're fine with changing the interpretation, I do like the new way better.

Reply:
Yeah, I think that it's the only semi-coherent way to make sense of these cases.